src/term/check/type_check.rs

```diff
@@ -0,0 +1,70 @@
+use core::fmt;
+use std::collections::HashMap;
+
+use crate::term::{Book, DefId, Name, RulePat};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Type {
+  Any,
+  Adt(Name),
+}
+
+pub type DefinitionTypes = HashMap<DefId, Vec<Type>>;
+
+impl Book {
+  /// Returns a HashMap from the definition id to the inferred pattern types
+  /// and checks the rules arities based on the first rule arity.
+  pub fn type_check(&self) -> anyhow::Result<DefinitionTypes> {
```

Comment:
This is more than just type checking. It checks the type and then returns it. Maybe `infer_types` or `infer_def_types` would be better, since what we do is infer the types from the used constructors and return an error in case the inference fails.